src/core/algorithms/gfd/gfd_miner.cpp

```diff
@@ -0,0 +1,648 @@
+
+    bool operator()(vertex_t const& fr, vertex_t const& to) const {
+        return query[fr].attributes.at("label") == graph[to].attributes.at("label");
+    }
+};
+
+struct ECompare {
+    graph_t const& query;
+    graph_t const& graph;
+
+    bool operator()(edge_t const& fr, edge_t const& to) const {
+        return query[fr].label == graph[to].label;
+    }
+};
+
+void ChangeLiterals(std::vector<Literal>& literals, graph_t const& pattern, graph_t const& existed,
+                    Embedding const& iso) {
+    for (auto& [fstToken, sndToken] : literals) {
+        if (fstToken.first != -1) {
+            auto vertex = boost::vertex(fstToken.first, pattern);
+            fstToken.first = existed[iso.at(vertex)].node_id;
+        }
+        if (sndToken.first != -1) {
+            auto vertex = boost::vertex(sndToken.first, pattern);
+            sndToken.first = existed[iso.at(vertex)].node_id;
+        }
+    }
+}
+
+void UpdateExisted(bool& isoExists, graph_t const& new_pattern,
+                   std::vector<graph_t> const& new_patterns,
+                   std::vector<Rules>& new_forbidden_rules_set, Rules const& forbidden_rules) {
+    std::size_t index = 0;
+    for (auto const& existed : new_patterns) {
+        VCompare vcompare{new_pattern, existed};
+        ECompare ecompare{new_pattern, existed};
+        Embedding iso;
+        CompareCallback callback(new_pattern, iso, isoExists);
+        boost::vf2_subgraph_iso(new_pattern, existed, callback,
+                                get(boost::vertex_index, new_pattern),
+                                get(boost::vertex_index, existed),
+                                vertex_order_by_mult(new_pattern), ecompare, vcompare);
+        if (!isoExists) {
+            index++;
+            continue;
+        }
+        Rules new_forbidden_rules(forbidden_rules);
+        for (auto& [premises, conclusion] : new_forbidden_rules) {
+            ChangeLiterals(premises, new_pattern, existed, iso);
+            ChangeLiterals(conclusion, new_pattern, existed, iso);
+            bool rule_exists = false;
+            for (auto& [new_premises, new_conclusion] : new_forbidden_rules_set.at(index)) {
+                if (CompareLiteralSets(premises, new_premises) &&
+                    CompareLiteralSets(conclusion, new_conclusion)) {
+                    rule_exists = true;
+                    break;
+                }
+            }
+            if (!rule_exists) {
+                new_forbidden_rules_set.at(index).push_back({premises, conclusion});
+            }
+        }
+        break;
+    }
+}
+
+void AddEdge(std::set<std::string> const& edge_labels, graph_t const& graph, graph_t const& pattern,
+             Embeddings const& embeddings, Rules const& forbidden_rules,
+             std::vector<graph_t>& new_patterns, std::vector<Embeddings>& new_embeddings_set,
+             std::vector<Rules>& new_forbidden_rules_set) {
+    std::size_t vertices = boost::num_vertices(pattern);
```

Comment:
Bad naming
`num_vertices` is more informative and less missleading.
Applies to everywhere else